src/main/java/de/tum/in/www1/artemis/service/util/structureoraclegenerator/OracleGenerator.java

```diff
@@ -140,22 +143,22 @@ private static Map<JavaClass, JavaClass> generateSolutionToTemplateMapping(Path
         List<Path> solutionFiles = retrieveJavaSourceFiles(solutionProjectPath);
         log.debug("Template Java Files {}", templateFiles);
         log.debug("Solution Java Files {}", solutionFiles);
+
         var templateClasses = getClassesFromFiles(templateFiles);
         var solutionClasses = getClassesFromFiles(solutionFiles);
 
-        Map<JavaClass, JavaClass> solutionToTemplateMapping = new HashMap<>();
+        // Convert template classes into a map for quick lookup
+        Map<String, JavaClass> templateClassMap = templateClasses.stream()
+                .collect(Collectors.toMap(cls -> cls.getPackageName() + "." + cls.getSimpleName(), Function.identity(), (existing, replacement) -> existing)); // In case of name
+                                                                                                                                                               // conflicts, keep
+                                                                                                                                                               // the existing
 
+        // Map each solution class to its counterpart in the template
+        Map<JavaClass, JavaClass> solutionToTemplateMapping = new HashMap<>();
         for (JavaClass solutionClass : solutionClasses) {
-            // Put an empty template class as a default placeholder.
-            solutionToTemplateMapping.put(solutionClass, null);
-
-            for (JavaClass templateClass : templateClasses) {
-                if (solutionClass.getSimpleName().equals(templateClass.getSimpleName()) && templateClass.getPackageName().equals(solutionClass.getPackageName())) {
-                    // If a template class with the same name and package gets found, then replace the empty template with the real one.
-                    solutionToTemplateMapping.put(solutionClass, templateClass);
-                    break;
-                }
-            }
+            String qualifiedName = solutionClass.getPackageName() + "." + solutionClass.getSimpleName();
```

Comment:
```suggestion
            String qualifiedName = solutionClass.getName();
```
```suggestion
            String qualifiedName = solutionClass. getCanonicalName();
```
Why doesn't one of those work?